카카오 키워드 검색 API 연동 후 및 지역 셀 선택 시, 해당 지역의 중점으로 이동 구현

Minny27 opened the pull request:
## Motivation ⍰

- 카카오 키워드 검색 API 연동 구현을 위함

<br>

## Key Changes 🔑

- 선택된 아이템을 받아서 아이템의 위도, 경도를 중심으로 지도 이동
- 문자열을 확장해서 특정 문자열의 폰트, 색상 변경 구현
- 셀 클릭 시, 이전 화면(MapVC)에 선택한 item 전달
-  Keychain 클래스 생성 시, apiKey 설정 구현 -> 런타임 후의 보안을 위함
미리보기는 다음과 같습니다.

<img src="https://github.com/NAGAZA-Team/NAGAZA-iOS/assets/68800789/7e98793a-efd5-4a3d-812d-5b28d4e6cf0b" width=25%>

고민한 사안들
1. `APIKey`는 서버로 부터 입력받지 않고 실행 전에 저장을 해야하기 때문에 보안상 문제가 있을 수 있습니다.
런타임 시 호출되는 함수에서 `API Key`를 set하게 되면, 앱이 해킹되었을 때 `API`의 보안 위험할 수 있습니다.
이를 위해 Keychain 객체를 생성할 때 set하고 런타임 시에 keychain에 접근하도록 구현하는 것이 안전할 수 있습니다.
(remote에는 해당 keychain이 남아 있기 때문에 remote에도 안 남기고 싶으면 레포지토리의 접근 권한을 private으로 해야할 것 같습니다.)

코드는 다음과 같습니다.

```swift
final class Keychain {
    static let shared = Keychain()
    
    private init() {
        setApiKey() // 싱글톤 객체 생성 시, ApiKey 설정
    }
    
    enum TokenType: String {
        case apiKey // apiKey 추가
        case accessToken
        case refreshToken
    }

    { ... }
    
    private func setApiKey() {
        set(
            "eb31c26138364bf0d6d5fa78138e693d",
            forKey: .apiKey
        )
    }
}

```

2. 텍스트 필드로 검색한 텍스트를 검색 리스트에서 색칠해주기
처음에는 `Entity`에 해당 프로퍼티를 옵셔널로 선언하고 값을 변경해주는 방식으로 구현하려고 했습니다.
하지만 `Entity`에 변수가 있다는 점이 `Domain`은 변화를 최소화해야한다는 점에서 문제가 있다고 생각했습니다.
그래서 model과 검색 text 두 파라미터를 받아서 처리하도록 구현했습니다.

코드는 다음과 같습니다.
처음 생각
```swift
struct Place {
    let identifier = UUID()
    var text: String? = nil // 검색한 텍스트 추가
    let addressName: String
    let categoryGroupCode: String
    let categoryGroupName: String
    let categoryName: String
    let distance: String
    let id: String
    let phone: String
    let placeName: String
    let placeURL: String
    let roadAddressName: String
    let x: String
    let y: String
}

text = searchText.text
```
Domain의 변경이 생겨서 해당 방식은 채택 x

수정 방안
```swift
// MapSearchViewController
private func setDataSource() {
    dataSource = DataSource(tableView: tableView, cellProvider: { [weak self] tableView, indexPath, itemIdentifier in
        guard let self = self else { return UITableViewCell() }
        
        guard let cell = tableView.dequeueReusableCell(withIdentifier: SearchTableViewCell.identifier) as? SearchTableViewCell
        else { return UITableViewCell() }
        
        // item과 검색어를 함께 파라미터로 넘겨줌
        cell.bind(
            item: itemIdentifier,
            keyword: self.searchTextField.text
        )
        
        return cell
    })
}

// SearchTableViewCell
func bind(item: Place, keyword: String?) {
    titleLabel.attributedText = item.placeName.toAttributedString(
        with: keyword,
        font: UIFont.ngaSubTitle1B,
        color: NagazaAsset.Colors.mainOrange.color
    )
    subtitleLabel.attributedText = item.addressName.toAttributedString(with: keyword)
}
    
```

다른 더 좋은 방법이 있으면 말씀해주세요!

<br>

## To Reviewers 🙏🏻

- [ ] 키워드 검색이 잘 되는지 확인해주세요.
- [ ] 검색된 지역 리스트 중 특정 셀을 클릭했을 때, 해당 화면이 사라지고 MapVC에서 선택된 지역으로 중심이 변경되는 지 확인해주세요.

<br>

## Linked Issue 🔗

- [ ] #20 
- [x] #21

Minny27 committed:
Update: search 키워드 추가 -> 이름을 명확하게 하기 위함

Minny27 committed:
Update: DataMapping을 Request, Response로 파일을 분리

- 파일 이동 이유는 request, response가 많아져서 헷갈림
- feature 별로도 나누어야 할 거 같아서 depth가 너무 많아짐
- KeywordSearchRequest encodable 구조체 생성
- 파일 이름도 변경

Minny27 committed:
[#21]Feat: 카카오 API 연동해서 키워드 검색을 위한 useCase, repository 구현

- url이 달라서 APITarget 새롭게 추가
- Keychain 클래스 생성 시, apiKey 설정 구현 -> 런타임 후의 보안을 위함
- 추후에 최근 검색 repository 구현할 예정

Minny27 committed:
[#31]Update: 검색 후, 응답받은 모델의 entity 생성

Minny27 committed:
Update: NagazaBaseTableViewController -> NagazaBaseTableView로 변경

- NagazaTableViewCell 추가

Minny27 committed:
[#31]Feat: 셀 클릭 시, 이전 화면(MapVC)에 선택한 item 전달

Minny27 committed:
[#31]Feat: SearchTableViewCell 구현

- 문자열을 확장해서 특정 문자열의 폰트, 색상 변경 구현
- 파일 위치 변경

Minny27 committed:
[#31]Feat: 선택된 아이템을 받아서 아이템의 위도, 경도를 중심으로 지도 이동

- view model 변수로 변경
- view model에 searchItem 추가
- 중점으로 이동하는 함수 구현

Minny27 committed:
[#31]Feat: 검색 API 연동 및 셀 선택 이벤트 구현

- collection view에서 tableView로 전한
- 이유는 separator를 커스텀해줘야 하고 복잡한 layout이 아니기 때문
- 검색 버튼을 눌렀을 때, 최근 리스트와 합성해서 리스트 중 선택된 index의 아이템을 MapVC에 전달

Minny27 committed:
Refactor: mapButtonTapSubject bindViewModel 함수에 포함

- mapButtonItem을 View Controller에 선언
- map button 이벤트를 bind view model에 적용

hililyy commented on Nagaza/Sources/Presentation/Feature/TabBar/Map/View/Cell/SearchTableViewCell.swift:
```diff
@@ -0,0 +1,58 @@
+final class SearchTableViewCell: NagazaTableViewCell {
+    static let identifier = SearchTableViewCell.description()
+
+    lazy var labelStackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.spacing = 6
+        stackView.alignment = .leading
+        return stackView
+    }()
+    
+    lazy var titleLabel: UILabel = {
+        let label = UILabel()
+        label.font = UIFont.ngaSubTitle1R
+        label.textAlignment = .center
+        label.textColor = NagazaAsset.Colors.black1.color
+        return label
+    }()
+    
+    lazy var subtitleLabel: UILabel = {
+        let label = UILabel()
+        label.font = UIFont.ngaP1R
+        label.textAlignment = .center
+        label.textColor = NagazaAsset.Colors.gray3.color
+        return label
+    }()
```
전에 의견주셨던 UI컴포넌트에 lazy를 붙이자고 하신 부분 늦었지만 고민해봤는데요
제가 생각할 때는 해당 컴포넌트들이 어차피 init 시점에 메모리가 바로 할당되는데 lazy로 하는게 맞을까 싶습니다.
바로 메모리가 할당되지만 그래도 그 잠깐의 순간이라도 메모리를 아낄 수 있다고 해도 lazy가 thread safety하지 않다는 점에서 예상할 수 없는 사이드 이펙트를 발생시킬 수 있을 것 같다고 생각이 드는데 어떻게 생각하시나요..?
lazy로 했을 때 손해볼게 없다고 하셨는데 어떤 부분에서 그렇게 생각하셨는지 궁금합니다..ㅠ

Minny27 replied:
이전에는 lazy var를 사용하면 해당 객체의 멤버에 접근해야 메모리에 올라가서 `lazy var`로 쓰는 게 무조건 효율적이라고 생각했습니다.
그런데 멀티 스레드 환경에서 동시에 객체의 멤버에 접근하면 생성이 여러 번 될 수 있고, 이로 인해 데이터 무결성에 영향을 줄 거 같습니다.
객체의 용량이 크지 않으면 `let`으로 통일하는 게 멀티 스레드 환경에서 안전할 거 같습니다.
좋은 지적이셨던 거 같고 배워갑니다!

hililyy commented on Nagaza/Sources/Presentation/Utils/Keychain/Keychain.swift:
```diff
@@ -113,5 +115,12 @@ final class Keychain {
+    
+    private func setApiKey() {
+        set(
+            "eb31c26138364bf0d6d5fa78138e693d",
```
음 지금처럼 key를 remote에 올리면 레포를 private으로 해야할 것 같은데..
그런데 이 프로젝트로 나중에 포폴에 쓰고싶을 수 있지 않을까요..?
이건 좀 더 고민해보면 좋을 것 같습니다..!

Minny27 replied:
고민해보긴 했는데, .gitignore에 파일 추가하고 그 파일에서 가져오는 방식으로 일단 해놓겠습니다.
추후에 private으로 전환하면 그 때 다시 수정하겠습니다!

hililyy commented:
요청주셨던 동작은 잘 되는거 확인했습니다. 고생많으셨습니다!

Minny27 committed:
Merge branch 'develop' into Feature/kakao_keyword_search

Minny27 committed:
[#31]Feat: 네이버 맵 Client Key 추가 및 설정

- podfile 추가
- 키체인 init 시, api key 설정 코드 추가
- appDelegate에서 naver api key 설정 코드 추가
- .gitignore에 Pods, Podfile.lock, Storage 추가

Minny27 committed:
[#31]Enhance: 검색어가 길어질 경우 대응

- 텍스트 라벨의 trailing을 찾기 이미지의 왼쪽 16pt까지로 설정

Minny27 committed:
Update: api 이름 변경

Minny27 committed:
Refactor: lazy var -> let으로 변경

- thread safe하게 구현하기 위함

Minny27 committed:
[#31]Refactor: apple map -> naver map으로 마이그레이션

- 필요없는 코드 삭제
- searchItem 바인딩 변수 추가
- place에 대한 지역명과 색상 변경 함수 추가

Minny27 commented:
수정 사항은 다음과 같습니다.
1. API Key 하드 코딩된 코드 걷어냈습니다.
- 현재 레포지토리를 public으로 설정되어 있어서 어떻게 할 지 고민하다가
- Storage라는 파일에 api key를 넣고 .gitignore에 해당 파일을 제외하는 방식으로 일단 구현했습니다. => 빌드되려면 해당 파일이 필요합니다.

2. apple map을 naver map으로 전환했습니다.
- naver map은 pod을 써야해서 같은 버전으로 맞추는 게 좋아보여서 네이버 맵 최신 버전(3.18.0)으로 설정했습니다.

3. lazy var -> let 전환
thread safe 하게 객체를 생성 및 접근하기 위해서 let이 더 안전하다고 생각해서 변경했습니다.

빌드 과정은 다음과 같습니다.
1. git pull
2. Storage 파일을 Keychain 폴더에 추가(Keychain 폴더 우클릭 > Add Files to Nagaza 로 추가)
3. terminal에서 Nagaza 프로젝트 파일이 있는 경로로 이동
4. tuist generate
5. pod install
6. Use Version on Disk

Minny27 committed:
Update: 중복된 BaseTableViewCell 삭제

Minny27 commented:
선택한 지역의 글자수가 길 때 처리도 추가했습니다.

<img src="https://github.com/NAGAZA-Team/NAGAZA-iOS/assets/68800789/f5fd4236-2876-4d61-a097-23734685f6c4" width=25%>
